daemon/SecretsImpl/dataprotector.cpp

```diff
@@ -0,0 +1,271 @@
+    // count how many times each occours. Ideally we end up with 3 occourences
+    // on the first index, and 0 on the others.
+    QVector<int> occourences(fileContents.size());
+    for (const QByteArray &currentData : fileContents) {
+        for (int i = 0; i < fileContents.size(); i++) {
+            if (currentData == fileContents.at(i)) {
+                occourences[i]++;
+                break;
+            }
+        }
+    }
+
+    // Assume that the version that occours the most is the correct version.
+    int maxOccourenceIndex = 0;
+    for (int i = 0; i < occourences.size(); i++) {
+        if (occourences[i] > occourences[maxOccourenceIndex]) {
+            maxOccourenceIndex = i;
+        }
+    }
+    int minimumValidOccourences = fileContents.size() / 2 + 1;
+    if (occourences[maxOccourenceIndex] < minimumValidOccourences) {
+        qWarning(lcSailfishSecretsDaemon) << "Could not decide which file is correct, there is no majority.";
+        return Irretrievable;
+    }
+
+    m_data = fileContents[maxOccourenceIndex];
+    *result = m_data;
+
+    return Success;
+}
+
+DataProtector::Status DataProtector::putData(const QByteArray &bytes)
+{
+    QDir dir(m_path);
+
+    // Check if the main directory exists, create if not
+    if (!dir.exists()) {
+        if (!dir.mkpath(m_path)) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't create protected root directory when writing new data:" << m_path;
+            return ErrorCannotCreateDirectory;
+        }
+
+        qCDebug(lcSailfishSecretsDaemon) << "Protected root directory didn't exist, so putData assumes the data is empty.";
+    }
+
+    // Get list of old directories, these will be deleted at the end
+    QFileInfoList oldDirectories = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot);
+
+    // Create new data directory
+    QString dataDirName = QUuid::createUuid().toString().remove('{').remove('}');
+    QString dataDirPath = m_path + QStringLiteral("/") + dataDirName;
+    if (!dir.mkdir(dataDirPath)) {
+        qCWarning(lcSailfishSecretsDaemon) << "Can't create data directory when writing new data:" << dataDirPath;
+        return ErrorCannotCreateDirectory;
+    }
+
+    // Create redundant files in new data directory
+    QVector<QString> dataFilePaths = {
+        dataDirPath + QStringLiteral("/file0"),
+        dataDirPath + QStringLiteral("/file1"),
+        dataDirPath + QStringLiteral("/file2"),
+    };
+    QVector<QSharedPointer<QSaveFile> > files;
+    std::transform(dataFilePaths.begin(), dataFilePaths.end(), std::back_inserter(files), [](const QString &filePath) {
+        return QSharedPointer<QSaveFile>(new QSaveFile(filePath));
+    });
+
+    bool allOpen = std::all_of(files.begin(), files.end(), [](QSharedPointer<QSaveFile> &file) {
+        bool isOpen = file->open(QIODevice::WriteOnly);
+        if (!isOpen) {
+            qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName();
```

Comment:
Sorry to add comments late, if you still have time, it would be nice to log the error string, so it's easier to debug or understand the reason of the error:
```qCWarning(lcSailfishSecretsDaemon) << "Can't open file for writing:" << file->fileName() << file->errorString();```

Reply:
Sorry, you were literally 5 minutes late with this. But that's a nice catch and I'll add it next chance I get.

Reply:
Yep, thought about it at lunch time, but was busy to report it before… Anyway, it's not critic, and can definitively come with a later commit.